File: python/tests/test_pathnorm.py
from forensicator.pathnorm import normalize_path, is_case_insensitive_os


def test_collapse_slashes():
    assert normalize_path("a//b///c", os="linux") == "a/b/c"


def test_backslash_to_slash():
    assert normalize_path("a\\b\\c", os="linux") == "a/b/c"


def test_strip_leading_dot_slash():
    assert normalize_path("./a/b", os="linux") == "a/b"


def test_strip_trailing_slash():
    assert normalize_path("a/b/", os="linux") == "a/b"


def test_root_preserved():
    assert normalize_path("/", os="linux") == "/"


def test_case_fold_on_mac():
    assert normalize_path("Photos/2024", os="darwin") == "photos/2024"


def test_no_case_fold_on_linux():
    assert normalize_path("Photos/2024", os="linux") == "Photos/2024"


def test_case_fold_on_windows():
    assert normalize_path("Photos\\2024", os="MSWin32") == "photos/2024"


def test_is_case_insensitive_os():
    assert is_case_insensitive_os("darwin")
    assert is_case_insensitive_os("MSWin32")
    assert is_case_insensitive_os("win32")
    assert not is_case_insensitive_os("linux")

File: cpp/tests/test_pathnorm.cpp
#include "forensicator/pathnorm.hpp"

#include <cassert>
#include <iostream>
#include <string>

using forensicator::normalize_path;

static void eq(const std::string& got, const std::string& expected, const char* msg) {
    if (got != expected) {
        std::cerr << "FAIL " << msg << ": got \"" << got << "\" expected \"" << expected << "\"\n";
        std::abort();
    }
}

int main() {
    // Slash collapsing
    eq(normalize_path("a//b///c", "linux"), "a/b/c", "collapse slashes");
    // Backslash conversion
    eq(normalize_path("a\\b\\c", "linux"), "a/b/c", "backslashes");
    // Trailing slash
    eq(normalize_path("foo/bar/", "linux"), "foo/bar", "trailing slash");
    // Root preserved
    eq(normalize_path("/", "linux"), "/", "root");
    // Leading ./
    eq(normalize_path("./a/b", "linux"), "a/b", "leading ./");
    // Case-insensitive on darwin
    eq(normalize_path("Foo/BAR.TXT", "darwin"), "foo/bar.txt", "darwin lowercase");
    // Case-sensitive on linux
    eq(normalize_path("Foo/BAR.TXT", "linux"), "Foo/BAR.TXT", "linux case-sensitive");
    // Windows case-insensitive
    eq(normalize_path("C:\\Users\\Bob", "MSWin32"), "c:/users/bob", "windows");
    // Combination
    eq(normalize_path("././/a//B/c\\d/", "darwin"), "a/b/c/d", "combo");
    std::cout << "test_pathnorm ok\n";
    return 0;
}
